composer/datasets/in_context_learning_evaluation.py

```diff
@@ -109,12 +198,193 @@ def update_metric(self, metric, batch, output_logits, labels):
+
+            encoded_example = {}
+            query, choices, gold_idx = self.samples[sample_idx]['query'], self.samples[sample_idx][
+                'choices'], self.samples[sample_idx]['gold'],
+            if len(preamble) > 0:
+                query = f'{example_delimiter}{query}'
+
+            choices = [f'{continuation_delimiter}{choice}' for choice in choices]
+            encoded_example['query'] = self.tokenizer(query)
+            encoded_example['choices'] = [self.tokenizer(choice) for choice in choices]
+            encoded_example['preamble'] = self.tokenizer(
+                preamble)  # if the preamble is empty then these will be 0-length lists
+            encoded_example['gold_idx'] = gold_idx
+
+            examples.append(encoded_example)
+
+        return examples
+
+    def __getitem__(self, index):
+        return self.encoded_dataset[index]
+
+    def __len__(self):
+        return len(self.encoded_dataset)
+
+    def collate_fn(self, data):
+        inputs = []
+        continuation_indices = []
+        gold_idxs = []
+        choice_groupings = []
+        for data_pair in data:
+
+            choice_start_idx = len(continuation_indices)
+            preamble, context, choices, gold_idx = (data_pair['preamble'], data_pair['query'], data_pair['choices'],
+                                                    data_pair['gold_idx'])
+
+            for choice in choices:
+                context_enc = preamble['input_ids'] + context['input_ids']
+                continuation_enc = choice['input_ids']
+                inp, continuation_span = _make_padded_input(context_enc, continuation_enc, self.max_seq_len,
+                                                            self.eos_tok_id)
+
+                inputs.append(inp)
+                continuation_indices.append(continuation_span)
+
+            gold_idxs.append(gold_idx)
+            choice_end_idx = len(continuation_indices)
+            choice_groupings.append((choice_start_idx, choice_end_idx))
+
+        # We run each distinct query + answer choice through the model separately and determine which
+        # answer has the lowest per-token-perplexity.
+        #
+        # If each question has N possible choices, all N must be grouped together as distinct elements of the batch
+        # since the batch may consist of multiple questions, the choice_groupings indicates
+        # which contiguous sequences of elements in the batch correspond to which question
+        # gold_indices indicates which of the [0, N-1] choices is the correct one for each question.
+        batch = {
+            'input_ids': torch.stack(inputs),
+            'continuation_indices': continuation_indices,
+            'mode': 'icl_task',
+            'labels': torch.stack(inputs),
+            'gold_indices': gold_idxs,
+            'choice_groupings': choice_groupings
+        }
+        batch['attention_mask'] = ~(batch['input_ids'] == self.eos_tok_id)
```

Comment:
I don't think this is quite right, like I think `attention_mask` is meant to be `1` where there are tokens and `0` where there is padding. I think the `eos_token_id` is an artifact of the GPT2Tokenizer but not a general thing.

I think the safest thing todo here is to copy what HF's default data collator does (maybe even import and use directly!): https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/data/data_collator.py#L1152-L1193